62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,782 @@
+                auto splitCurveMin = left.getSplitCurve();
+                auto splitCurveMax = right.getSplitCurve();
+				assert(splitCurveMin.evaluateBezier(0.0)[major] <= splitCurveMin.evaluateBezier(1.0)[major]);
+				assert(splitCurveMax.evaluateBezier(0.0)[major] <= splitCurveMax.evaluateBezier(1.0)[major]);
+
+                auto curveMinAabb = splitCurveMin.getBezierBoundingBoxMinor();
+                auto curveMaxAabb = splitCurveMax.getBezierBoundingBoxMinor();
+                curveBox.aabbMin = float64_t2(std::min(curveMinAabb.first.x, curveMaxAabb.first.x), lastMajor);
+                curveBox.aabbMax = float64_t2(std::max(curveMinAabb.second.x, curveMaxAabb.second.x), newMajor);
+
+				if (isCurrentDebugStep)
+				{
+					drawDebugBezier(splitCurveMin, float64_t4(1.0, 0.0, 0.0, 1.0));
+					drawDebugBezier(splitCurveMax, float64_t4(0.0, 1.0, 0.0, 1.0));
+
+					//printf(std::format("Split curve min ({}, {}), ({}, {}), ({}, {}) max ({}, {}), ({}, {}), ({}, {})\n",
+					//	splitCurveMin.p[0].x, splitCurveMin.p[0].y,
+					//	splitCurveMin.p[1].x, splitCurveMin.p[1].y,
+					//	splitCurveMin.p[2].x, splitCurveMin.p[2].y,
+					//	splitCurveMax.p[0].x, splitCurveMax.p[0].y,
+					//	splitCurveMax.p[1].x, splitCurveMax.p[1].y,
+					//	splitCurveMax.p[2].x, splitCurveMax.p[2].y
+					//).c_str());
+
+				}
+
+                //if (debugOutput)
+                //{
+                //    drawDebugLine(float64_t2(curveBox.aabbMin.x, curveBox.aabbMin.y), float64_t2(curveBox.aabbMax.x, curveBox.aabbMin.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //    drawDebugLine(float64_t2(curveBox.aabbMax.x, curveBox.aabbMin.y), float64_t2(curveBox.aabbMax.x, curveBox.aabbMax.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //    drawDebugLine(float64_t2(curveBox.aabbMin.x, curveBox.aabbMax.y), float64_t2(curveBox.aabbMax.x, curveBox.aabbMax.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //    drawDebugLine(float64_t2(curveBox.aabbMin.x, curveBox.aabbMin.y), float64_t2(curveBox.aabbMin.x, curveBox.aabbMax.y), float32_t4(0.0, 0.3, 0.0, 0.1));
+                //}
+                // Transform curves into AABB UV space and turn them into quadratic coefficients
+                // TODO: the split curve should already have the quadratic bezier as
+                // quadratic coefficients
+                // so we wont need to convert here
+                auto transformCurves = [](Hatch::QuadraticBezier bezier, float64_t2 aabbMin, float64_t2 aabbMax, float64_t2* output) {
+                    auto rcpAabbExtents = float64_t2(1.0,1.0) / (aabbMax - aabbMin);
+                    auto p0 = (bezier.p[0] - aabbMin) * rcpAabbExtents;
+                    auto p1 = (bezier.p[1] - aabbMin) * rcpAabbExtents;
+                    auto p2 = (bezier.p[2] - aabbMin) * rcpAabbExtents;
+                    output[0] = p0 - 2.0 * p1 + p2;
+					if (output[0].y < 1e-6) output[0].y = 0.0;
```

Comment:
I did it differently here since we can do it in the hatch box AABB space, making it more accurate and not reliant in world space coordinates

Reply:
Ok, keep in mind that I also asked for some edits in `isLineSegment` which does relative comparison of dot(A,A) and dot(B,B).
and since it would be relative comparison, it shouldn't depend on the space you're in and you can choose your threshold as you see fit. It's exp2(-23) in closest distance function in beziers.hlsl but you can make it a lil higher for the purpose here to something like 1e-6

you're free to keep this if this feels safe to you, just assert here that the line `output[1]/B`.major isn't zero otherwise you got yourself a constant line in major direction which shouldn't go through the algo